test/tools/router_check/router.cc

```diff
@@ -0,0 +1,211 @@
+#include "test/tools/router_check/router.h"
+
+Json::ObjectPtr RouterCheckTool::loadJson(const std::string& config_json,
+                                          const std::string& schema) {
+
+  try {
+    // Json configuration
+    Json::ObjectPtr loader = Json::Factory::LoadFromFile(config_json);
+    loader->validateSchema(schema);
+    return loader;
+
+  } catch (const EnvoyException& ex) {
+    std::cerr << "config schema JSON load failed: " << config_json << std::endl;
+    std::cerr << ex.what() << std::endl;
+    return nullptr;
+  }
+}
+
+bool RouterCheckTool::initializeFromConfig(const std::string& router_config_json) {
+
+  Json::ObjectPtr loader = loadJson(router_config_json, Json::Schema::ROUTE_CONFIGURATION_SCHEMA);
+
+  if (loader != nullptr) {
+    config_.reset(new Router::ConfigImpl(*loader, runtime_, cm_, false));
+    return true;
+  }
+  return false;
+}
+
+bool RouterCheckTool::compareEntriesInJson(const std::string& expected_route_json) {
+  // Load tool config json
+  Json::ObjectPtr loader = loadJson(expected_route_json, Json::ToolSchema::routerCheckSchema());
+
+  if (loader == nullptr) {
+    return false;
+  }
+
+  bool no_failures = true;
+  // Iterate through each test case
+  for (const Json::ObjectPtr& check_config : loader->asObjectArray()) {
+    // Load parameters from json
+    ToolConfig tool_config;
+    tool_config.parseFromJson(check_config);
+    Router::RouteConstSharedPtr route =
+        config_->route(tool_config.headers_, tool_config.random_value_);
+
+    Json::ObjectPtr check_type = check_config->getObject("check");
+
+    // Call appropriate function for each match case
+    if (check_type->hasObject("path_redirect")) {
+      if (!compareRedirectPath(tool_config, check_type->getString("path_redirect"), route)) {
+        no_failures = false;
```

Comment:
This is good advice :) @htuch beat me to it. Instead of the long sequence of ifs you can write something like:

```c++
for (auto kv : function_map) {
  if (check_type->hasObject(kv.first)) {
    auto test_result = kv.second(tool_config, check_type->getString(kv.first), route); // assuming you don't want to early exit
    no_failures = no_failures ? test_result : false;
  }
}
```

Or iterate over found objects in JSON and look up the function in the map in a similar fashion.